### ISSUE TITLE
[bug] TypeError: Cannot read property 'name' of undefined

### ISSUE THREAD

**Issue by lookfirst** [mikro-orm/mikro-orm#153]
I'm trying to debug some weird error where mocha runs fine in my IDE, but fails on the command line. This is a minor issue that can be dealt with later.

```
     TypeError: Cannot read property 'name' of undefined
      at Function.fromMergeWithoutPK (node_modules/mikro-orm/dist/utils/ValidationError.js:50:68)
      at EntityValidator.validatePrimaryKey (node_modules/mikro-orm/dist/entity/EntityValidator.js:55:43)
      at EntityManager.merge (node_modules/mikro-orm/dist/EntityManager.js:141:24)
      at UnitOfWork.commitChangeSet (node_modules/mikro-orm/dist/unit-of-work/UnitOfWork.js:190:25)
      at Function.runSerial (node_modules/mikro-orm/dist/utils/Utils.js:189:22)
      at promise (node_modules/mikro-orm/dist/unit-of-work/UnitOfWork.js:80:39)
      at /functions/node_modules/mikro-orm/dist/connections/AbstractSqlConnection.js:29:29
```

```
    static fromMergeWithoutPK(meta) {
        throw new ValidationError(`You cannot merge entity '${meta.name}' without identifier!`);
    }

```

It seems like somehow meta is getting passed into the function as undefined. It might be helpful to add some protections higher up in the call stack, around doing that.

**Comment by B4nan** [comment]
In pretty much all cases this is about metadata discovery. I keep adding more and more validation there... 

What is the code that produces the error? It might not be discovery problem, but just a badly propagated entity name.

**Comment by lookfirst** [comment]
Yea, I think just bad config. I got the config working better, so no more errors. Before I was messing with baseUrl and that was causing a mess between mocha and runtime. Now I just always use cwd() of the project.

**Comment by B4nan** [comment]
Can we close this? I added validation to `MetadataStorage.get()` so it will throw error right about missing metadata for given entity, so at least it will not crash like this (not yet published, will be part of `alpha.13`). 

There is also validation for no entities discovered (which is usually the root cause of this), that should be triggered before this whole thing, its strange that it was not thrown for you here...

**Comment by lookfirst** [comment]
I see what's happening I think.... if I screw up the config and somehow Mikro starts to process my built javascript files instead of my ts files, then it gets all unhappy further down the road and outputs errors similar to the stuff above.

The reason my config gets screwed up is because I'm dealing with setting things up for both running from mocha and running for real. In this case, mocha is getting wrong paths. I can figure that part out on my own, but I think what would be good is to fail when processing js files. Which sets up another conundrum... I know you can't do that since you want to support the JS usecase too...

```
[discovery] ORM entity discovery started
[discovery] - processing 6 files from directory ./build/src/db/model
[discovery] - processing entity Dog
[discovery] - ignoring file Dog.js.map
[discovery] - processing entity Person
[discovery] - ignoring file Person.js.map
[discovery] - processing entity Thing
[discovery] - ignoring file Thing.js.map
[discovery] - entity discovery finished after 3 ms
```

**Comment by lookfirst** [comment]
I see the problem now. I have to force `tsNode: true` in the mikro options when running via mocha without specifying `mocha --require ts-node/register` at runtime. Just running `mocha` isn't enough. Regardless of what is in the mocha configuration file...

```
process.env.NODE_ENV = 'test';

module.exports = {
  "extension": [
    "ts"
  ],
  "spec": "test/**/*.spec.ts",
  "require": "ts-node/register"
};
```

Basically, I think the detection here is to see if ts-node/typescript is installed and if so, throw a huge warning or error that you think it is a TS project, but you're processing the JS files.

**Comment by lookfirst** [comment]
Also, even if I run with `mocha --require ts-node/register`, printing out the `orm` object created by `.init()`, it still shows `tsNode: false`. I think that internally there should be some switch to force that to be true if you detect this is a TS project.

**Comment by B4nan** [comment]
> Also, even if I run with mocha --require ts-node/register, printing out the orm object created by .init(), it still shows tsNode: false. I think that internally there should be some switch to force that to be true if you detect this is a TS project.

It will show `tsNode: false` because that was passed in the configuration (as default value), it is not changed. This is the code for auto-detection:

    const tsNode = this.config.get('tsNode') || process.argv[0].endsWith('ts-node') || process.argv.slice(1).some(arg => arg.includes('ts-node'));

And it is used only to look up entities based on correct folder (`entitiesDirs` vs `entitiesDirsTs`). 

> The reason my config gets screwed up is because I'm dealing with setting things up for both running from mocha and running for real. In this case, mocha is getting wrong paths. I can figure that part out on my own, but I think what would be good is to fail when processing js files. Which sets up another conundrum... I know you can't do that since you want to support the JS usecase too...

Well it should still work as expected. The whole purpose of `tsNode` flag in config is to not look into `entitiesDirs` when running via `ts-node`. You should always point your compiled entities into `entitiesDirs` and your source entities into `entitiesDirsTs`. ORM will look into correct folder based on the autodetection. 

Obviously when there is nothing in the arguments (like `mocha --require ts-node/register`) that can be used for autodetection, you will need to either handle different ORM configuration regarding `entitiesDirs` vs `entitiesDirsTs` or manually pass the `tsNode` option. 

I don't want to do auto detection for stuff like `mocha` configured in some external config file (`.mocharc.js` here), I think it should be user's responsibility to configure the ORM for tests via `tsNode` flag in this case. Sure it is definitely something that should be documented, probably in the installation section.

**Comment by B4nan** [comment]
Hey! Just found out how to do the autodetection better, finally spotted where does `ts-node` change the global space:

    require.extensions['.ts'] // this is undefined if ts-node is not registered

Will improve the autodetection later today, then it should work without changing the config.

**Comment by B4nan** [comment]
Available in `3.0.0-alpha.14`
